### PR TITLE
[3.3.5] Core/Spell: Chain Targeting Distance

### PR DESCRIPTION
**Pull request by HelloKitty**
Firstly, there were a couple of ways to implement this. This seemed like the simplest though.

**Changes proposed:**

-  Added WorldObject methods for calculating distance from the center
-  Added new target check predicate for target searching for spells from the center
-  Change how chain targets are found using the new predicate and distance calculations
- Change overall change behavior to what I believe are expected behavior fixing issues like extended C'thun beam chain and Thorim chain lightning issues.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)

Couldn't find a specific issue on TC's tracker. This is essentially an issue for range for things like chain lightning or chain heal. For example these spells are usually chaining 11.5 yards, the 1.5 being a result of the default UNIT_COMBAT_RANGE, and this can be exploited in many ways by increasing object scale. Under bloodlust and other size increasing effects chain heal can gain a considerable advantage. During mechanics that involving chaining things like Noggen Fogger or Baby Spice and reduce the range but of course not lower than the original intended distance.

This is basically an issue of chain target selection distance.

**Tests performed:** (Does it build, tested in-game, etc.)

Go to Gurubashi, easy place to test this without worrying about faction. Use a chain lightning spell. For example spell ID 10605.

Without change players at:
.go xyz -13227 256 22
.go xyz -13215.5 256 22

Are within chain potential for change lightning. This is exactly 11.5 yards. The 1.5y increase over the 10 yard expectation comes from the UNIT_COMBAT_REACH of the player's at the moment. Larger players under the effect of object scale increasing effects will be chained further away.

With change players at:
.go xyz -13227 256 22
.go xyz -13216.9 256 22

Do **NOT** chain any longer. You must be within 10 yards with this change.

With change players at:
.go xyz -13227 256 22
.go xyz -13217.01 256 22

Do chain meaning it doesn't break chaining behavior.

**Known issues and TODO list:** (add/remove lines as needed)

I don't know the scope of spells/mechanics that depend on this. Hopefully someone else does. As of right now I'm unaware of issues this causes.

Here is a gif of it chaining between 3 targets all exactly +/- 0.01m due meters which I can attribute only to height of terrain or maybe differences in center for the models?

http://imgur.com/kf8WunO

### CODE REVIEW

File: src/server/game/Entities/Object/Object.cpp [hunk 1]
```diff
@@ -1213,11 +1213,28 @@ bool WorldObject::IsWithinDist2d(const Position* pos, float dist) const
     return IsInDist2d(pos, dist + GetCombatReach());
 }
 
+bool WorldObject::IsWithinDist2dCenter(float x, float y, float dist) const
```
Comment (Shauren) [inline src/server/game/Entities/Object/Object.cpp]: imo just use IsInDist2d directly instead of this strangely named method

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Object/Object.cpp (hunk 1 above)
Comment (HelloKitty) [inline src/server/game/Entities/Object/Object.cpp]: Yea, that's reasonable. I only added it to fit the format of the currently used IsWithinDist2d. If you're ok with directly using that method that's fine with me. I will make that change.

**Comment by HelloKitty** [comment]
There is interesting [behavior discussion](https://us.battle.net/forums/en/wow/topic/16201251069) mentioned in this thread but no way to verify the validity since it's just randoms on the WoW forums. Additionally no way to determine if it is relevant for 3.3.5 either. 

I'm open to changing this PR further to cover any other behaviors of chain targeting that aren't behaving as they did on retail.

**Comment by Treeston** [comment]
No need to add a new method to `WorldObject`, just add optional params (same as `IsWithinDistInMap`) that are forwarded.

**Comment by HelloKitty** [comment]
@Treeston Seems reasonable, it's been a long time since I've looked at this so I can't see why I hadn't originally done that. Seems odd that I would have missed this.

**Comment by tkrokli** [comment]
@HelloKitty : Have you got time to check those 2 conflicting lines in Spell.cpp and sync your changes with current source?
- https://github.com/TrinityCore/TrinityCore/pull/19412/files#diff-952c711803610aa78110e4f92ff2516aR1862
- https://github.com/TrinityCore/TrinityCore/pull/19412/files#diff-952c711803610aa78110e4f92ff2516aR1877

**Comment by HelloKitty** [comment]
@tkrokli I do not, this was from a long while back and it can probably be implemented in a better way. More information is still needed about the expected behavior too. This was mostly a fix for chaining spells chaining further than they should and scaling based on the object's size which made fights like Thorim significantly more difficult.

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Spells/Spell.cpp

**Comment by jackpoz** [comment]
@HelloKitty do you have time to wrap up this PR so it can be merged ?

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -7782,6 +7793,30 @@ bool WorldObjectSpellAreaTargetCheck::operator()(WorldObject* target)
     return WorldObjectSpellTargetCheck::operator ()(target);
 }
 
+WorldObjectSpellChainTargetCheck::WorldObjectSpellChainTargetCheck(float range, Position const* position, Unit* caster,
+    Unit* referer, SpellInfo const* spellInfo, SpellTargetCheckTypes selectionType, ConditionContainer* condList)
+    : WorldObjectSpellTargetCheck(caster, referer, spellInfo, selectionType, condList), _range(range), _position(position) { }
+
+bool WorldObjectSpellChainTargetCheck::operator()(WorldObject* target)
+{
+    //TODO: Add center range for gameobject targeting for chaining
```
Comment (jackpoz) [inline src/server/game/Spells/Spell.cpp]: is this TODO a blocking issue or can it be merged anyway ?

**Comment by Treeston** [comment]
Rebase, don't merge. @tkrokli

**Comment by tkrokli** [comment]
OK, I see. I used the online editor, which only defaults to "merge", but I also did my best to ensure that only the added changes were kept from this PR where it conflicted with the current 3.3.5 version of the file `src/server/game/Spells/Spell.cpp` (2 lines). I presume you recommend that it would have been better to omit the changes from the 3.3.5 branch in those 2 lines in this PR when I solved the file conflict. If that is the case, I can remove the 3.3.5 addition which is highlighted in dark green in the merge commit (if that is what you recommend).

**Comment by Treeston** [comment]
Don't merge at all. Rebase the branch on current 3.3.5 head.

PS:
> Or just don't touch peoples' branches without good reason.

**Comment by tkrokli** [comment]
Very well. To me, it looked like it would take quite some time before @HelloKitty would reply and update this PR, seeing that Aokromes pointed out the conflict already at 2017-07-18 and without any reaction so far. That is why I chose to solve the file conflict to prepare it for merging, also seeing that jackpoz asked for updates and feedback.

**Comment by HelloKitty** [comment]
@tkrokli Hi, I'm unavailable to work on this. I think more information is still needed as this would effect player chain lightning jump distances. Which, if I recall, was disputed in the forum post I linked to. Unsure the behavior it should have. Also could be implemented better probably, was something hacked together real quick to fix an issue as fast as possible. I wish I could be more helpful but I am too busy with other projects at this time to take a look again at this PR.

**Comment by tkrokli** [comment]
Very well. Thanks for the feedback anyway.
So it will be up to main TC devs to decide whether to keep or dismiss this PR.

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Spells/Spell.cpp
src/server/game/Spells/Spell.h

**Comment by tkrokli** [comment]
`...` possibly label this PR as [Closed-Abandoned] ?

**Comment by jackpoz** [comment]
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

**Pull request closed without merging**